### PR TITLE
mtls refactor for endpoints

### PR DESCRIPTION
**Pull request by ScottSuarez**
added support for mtls authentication 

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
added support for [mtls authentication](https://google.aip.dev/auth/4114) 
```

ref: go/dca:terraform

### CODE REVIEW

File: mmv1/third_party/terraform/utils/mtls_util.go [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+func getMtlsEndpoint(baseEndpoint string) string {
+	u, err := url.Parse(baseEndpoint)
+	if err != nil {
+		if strings.Contains(baseEndpoint, ".googleapis") {
+			return strings.Replace(baseEndpoint, ".googleapis", ".mtls.googleapis", 1)
+		}
+		return baseEndpoint
+	}
+	portParts := strings.Split(u.Host, ":")
```
Comment (slevenick) [inline mmv1/third_party/terraform/utils/mtls_util.go:35]: Do we have any endpoints that include the port in this format? This code is only run on our default base paths, so why would we need to include this?

Comment (ScottSuarez) [reply mmv1/third_party/terraform/utils/mtls_util.go:35]: this is a forked version of the function from https://github.com/GoogleCloudPlatform/declarative-resource-client-library/blob/main/dcl/transport.go#L234 .. I can revise it to be more specific to our scenario ?

File: mmv1/third_party/terraform/utils/mtls_util.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	"fmt"
+	"net/url"
+	"strings"
+
+	"google.golang.org/api/option/internaloption"
+	"google.golang.org/api/transport"
+)
+
+func isMtls() bool {
+	regularEndpoint := "https://mockservice.googleapis.com/v1/"
```
Comment (shinfan) [inline mmv1/third_party/terraform/utils/mtls_util.go:14]: Please add some comments to explain the reason behind this logic

File: mmv1/third_party/terraform/utils/mtls_util.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+	mtlsEndpoint := getMtlsEndpoint(regularEndpoint)
+	_, endpoint, err := transport.NewHTTPClient(context.Background(),
+		internaloption.WithDefaultEndpoint(regularEndpoint),
+		internaloption.WithDefaultMTLSEndpoint(mtlsEndpoint),
+	)
+	if err != nil {
+		return false
+	}
+	isMtls := strings.Contains(endpoint, "mtls")
```
Comment (shinfan) [inline mmv1/third_party/terraform/utils/mtls_util.go:23]: endpoint == mtlsEndpoint should be enough

**Review by shinfan — APPROVED** [review]

File: mmv1/third_party/terraform/utils/provider_handwritten_endpoint.go.erb [hunk 4]
```diff
@@ -151,40 +136,27 @@ var ServiceUsageCustomEndpointEntry = &schema.Schema{
...
 var BigtableAdminCustomEndpointEntry = &schema.Schema{
 	Type:         schema.TypeString,
 	Optional:     true,
 	ValidateFunc: validateCustomEndpoint,
 	DefaultFunc: schema.MultiEnvDefaultFunc([]string{
 		"GOOGLE_BIGTABLE_CUSTOM_ENDPOINT",
-	}, BigtableAdminDefaultBasePath),
-}
-
-var EventarcDefaultBasePath = "https://eventarc.googleapis.com/v1/"
```
Comment (slevenick) [inline mmv1/third_party/terraform/utils/provider_handwritten_endpoint.go.erb:179]: Lets not remove this yet, I have a PR to fix some generated base path issues coming

**Review by slevenick — COMMENTED** [review]
How will this work for DCL-based resources?

File: mmv1/third_party/terraform/utils/provider_handwritten_endpoint.go.erb (hunk 4 above)
Comment (ScottSuarez) [reply mmv1/third_party/terraform/utils/provider_handwritten_endpoint.go.erb:179]: left in after discussion

File: mmv1/third_party/terraform/utils/config.go.erb [hunk 5]
```diff
@@ -128,10 +128,51 @@ type Config struct {
...
+	ContainerBetaBasePathKey : "https://container.googleapis.com/v1beta1/",
+	DataprocBetaBasePathKey : "https://dataproc.googleapis.com/v1beta2/",
+	DataflowBasePathKey : "https://dataflow.googleapis.com/v1b3/",
+	IAMBasePathKey : "https://iam.googleapis.com/v1/",
+	IamCredentialsBasePathKey : "https://iamcredentials.googleapis.com/v1/",
+	ResourceManagerV2BasePathKey : "https://cloudresourcemanager.googleapis.com/v2/",
+	ServiceNetworkingBasePathKey : "https://servicenetworking.googleapis.com/v1/",
+	StorageTransferBasePathKey : "https://storagetransfer.googleapis.com/v1/",
+	BigtableAdminBasePathKey : "https://bigtableadmin.googleapis.com/v2/",
+	EventarcBasePathKey : "https://eventarc.googleapis.com/v1beta1/",
```
Comment (slevenick) [inline mmv1/third_party/terraform/utils/config.go.erb]: This should switch based on the version we are compiling at

**Review by slevenick — APPROVED** [review]
LGTM after fix for eventarc base path

File: mmv1/third_party/terraform/utils/config.go.erb (hunk 5 above)
Comment (ScottSuarez) [reply mmv1/third_party/terraform/utils/config.go.erb]: this key isn't used... dcl actually internally hardcodes their own basepath. I remove this reference !

**Comment by ScottSuarez** [comment]
test passed on previous run.. merging

**Pull request merged**